components/retro-go/drivers/audio/i2s.c

```diff
@@ -123,12 +173,22 @@ static bool driver_submit(const rg_audio_frame_t *frames, size_t count)
     size_t pos = 0;
 
     // In speaker mode we use left and right as a differential mono output to increase resolution.
-    bool differential = state.device == 0;
+    // bool differential = state.device == 0;
+    bool differential = 0;
 
     for (size_t i = 0; i < count; ++i)
     {
-        int left = frames[i].left * volume;
-        int right = frames[i].right * volume;
+        int left = (frames[i].left * volume) + 32768;
+        int right = (frames[i].right * volume) + 32768;
```

Comment:
Yes so here's where it gets tricky, and possibly wrong :-)  

First of all, we're talking about the internal DAC of the ESP32.  The original ESP32 has two of those on io 25 and 26, but unfortunately they're only 8 bit.  The ESP32-S3 used in the 2024 badge doesn't have any internal DAC but the 2022 badge does. 

You're right, in my mind this code would only ever run for the internal DAC, but I now see it's the same code for an external DAC.  So at the very least we'll need a additional #if RG_AUDIO_USE_INT_DAC so as to leave the external DAC logic untouched.

Now for the reasoning.  We're focusing on the internal DAC.  This one is 8-bit. 

The i2s subsystem can output to the internal DAC and to me the driver initialisation code on lines 81-90 looks correct: it configures the system to accept 16 bit input (a sample for right and a sample for left each time) and to take only the most significant byte of every sample:
            .bits_per_sample = 16,
            .channel_format = I2S_CHANNEL_FMT_RIGHT_LEFT,
            .communication_format = I2S_COMM_FORMAT_STAND_MSB,

If the input would be in the range 0 to 64k, the above would be all we need.  But it's -32k to 32k in retro-go.  The addition of 32768 brings it in the required range of 0 to 64k.  An int on the ESP32 is 32 bit so it should be large enough.  The code could have been clearer on that, I basically kept the int that was already there.  Taking the MSB of every sample in the range 0 to 64k brings it in the range 0 to 256 for the DAC.

What happens next is that the regular retro-go code on line 218 and 219 puts the int for left and the int for right in the DMA buffer which is declared as a buffer of two times int16_t (one for left, one for right).  I am assuming that these assignments keep all the bits intact, but I'm not sure.  Line 223 then makes sure the memory buffer is written to the internal DAC through DMA.  Here I assume that the MSB conversion interprets the bits in the buffer as _unsigned_ 16 bit values and takes the MSB of every sample, which is what we want.

That's the theory, but it has too many assumptions.  Moreover, as I explained in the Fri3d Camp discord, when using the sine wave for testing, the output I'm seeing on the DAC is not a sine but half a sine.  The samples from 128 to 256 seem correct, but there are no values between 0 and 128, these all end up as 128 in the output signal.  So clearly there is something wrong.  Problem is, I'm not sure what.  Possibly it's the code, maybe it's a measuring mistake.

Interestingly, with the volume set at 5% in retro-go, I do measure a nice sine wave.  The volume at 25% results in some clipping on the lower part of the wave, and with the volume at 100% I see half the sine wave clipped as outlined above.

So this is why I don't regard the current code for the internal DAC to be final.  We should get a clean sine wave and we're not getting that.  I also don't understand why the original retro-go code doesn't add 32768.  I don't see how it can actually work without.

Note that with the current code, the sound is "reasonable" using the game-on, as long as you set the volume to 5%.  For my ears, that is more than loud enough.  But I would be happier if we could fully explain and fix the clipping.